src/gluonnlp/data/dataloader.py

```diff
@@ -19,45 +19,101 @@
 """DataLoader. An extension of Gluon data loader that allows multi-shard sampling."""
 __all__ = ['ShardedDataLoader']
 
-import sys
+import io
+import pickle
+import multiprocessing
+from multiprocessing.pool import ThreadPool
 from mxnet import context
-from mxnet.gluon.data.dataloader import DataLoader
-from mxnet.gluon.data.dataloader import _MultiWorkerIter, _as_in_context
-from mxnet.recordio import MXRecordIO
+from mxnet.gluon.data.dataloader import ForkingPickler, _as_in_context
+from mxnet.gluon.data.dataloader import default_mp_batchify_fn, default_batchify_fn
+from mxnet.gluon.data import sampler as _sampler
+
+_worker_dataset = None
+def _worker_initializer(dataset):
+    """Initialier for processing pool."""
+    # global dataset is per-process based and only available in worker processes
+    # this is only necessary to handle MXIndexedRecordIO because otherwise dataset
+    # can be passed as argument
+    global _worker_dataset
+    _worker_dataset = dataset
+
+def _worker_fn(samples, batchify_fn, dataset=None):
+    """Function for processing data in worker process."""
+    # pylint: disable=unused-argument
+    # it is required that each worker process has to fork a new MXIndexedRecordIO handle
+    # preserving dataset as global variable can save tons of overhead and is safe in new process
+    global _worker_dataset
+    if isinstance(samples[0], (list, tuple)):
+        batch = [batchify_fn([_worker_dataset[i] for i in shard]) for shard in samples]
+    else:
+        batch = batchify_fn([_worker_dataset[i] for i in samples])
+    buf = io.BytesIO()
+    ForkingPickler(buf, pickle.HIGHEST_PROTOCOL).dump(batch)
+    return buf.getvalue()
+
+def _thread_worker_fn(samples, batchify_fn, dataset):
+    """Threadpool worker function for processing data."""
+    if isinstance(samples[0], (list, tuple)):
+        batch = [batchify_fn([dataset[i] for i in shard]) for shard in samples]
+    else:
+        batch = batchify_fn([dataset[i] for i in samples])
+    return batch
+
+class _MultiWorkerIter(object):
```

Comment:
Why do we have to copy the full implementation?

Reply:
1.4.0 not yet released, better save the effort to move dependency forward to a nightly build

Reply:
Ok, I will approve it once it passes the ci test.